### ISSUE TITLE
Is there any way to improve the results?

### ISSUE THREAD

**Issue by hee100** [ultralytics/yolov5#6132]
### Search before asking

- [x] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I am making a model that distinguishes animals.
picture data is a cutout of the frame of the video. I finished various preprocessing, but I didn't get good results.
Is there any way to improve it?
And why did this result come out?

**train code**
`python train.py --img 640 --batch 144 --epochs 500 --cfg models/yolov5s.yaml --data data/custom2.yaml --weights yolov5s.pt --workers 16 --name final2_day --patience 300 --label-smoothing 0.05 --cache`

### train results
![labels](https://user-images.githubusercontent.com/95641633/147720138-d2fb65bb-7aed-4770-987c-c824da42d320.jpg)

![labels_correlogram](https://user-images.githubusercontent.com/95641633/147723424-67144404-f5a4-4326-a712-465228903207.jpg)

![confusion_matrix](https://user-images.githubusercontent.com/95641633/147720050-4b6eeb36-4e34-469c-868f-81231113c94c.png)

![F1_curve](https://user-images.githubusercontent.com/95641633/147720063-625b215d-71b4-49ed-9032-9a4dfb3eff25.png)

![P_curve](https://user-images.githubusercontent.com/95641633/147720081-8bdb8d89-6a24-44e6-964e-0c8932084d93.png)

![R_curve](https://user-images.githubusercontent.com/95641633/147720089-5560ed47-d7d1-4fd1-a432-cf03f19cb8f7.png)

![PR_curve](https://user-images.githubusercontent.com/95641633/147720087-a18534a3-b3c5-4e8f-b16c-aa9d0a11a1f7.png)

![results](https://user-images.githubusercontent.com/95641633/147720101-23c1d2e5-fd2e-4a2b-8875-98418ea6f75b.png)

**hyp**
`lr0: 0.01
lrf: 0.2
momentum: 0.937
weight_decay: 0.0005
warmup_epochs: 3.0
warmup_momentum: 0.8
warmup_bias_lr: 0.1
box: 0.05
cls: 0.5
cls_pw: 1.0
obj: 1.0
obj_pw: 1.0
iou_t: 0.2
anchor_t: 4.0
fl_gamma: 0.0
hsv_h: 0.015
hsv_s: 0.7
hsv_v: 0.4
degrees: 20.0
translate: 0.1
scale: 0.5
shear: 0.0
perspective: 0.0
flipud: 0.0
fliplr: 0.5
mosaic: 1.0
mixup: 0.0
copy_paste: 0.0
`

### val.py results  
![confusion_matrix](https://user-images.githubusercontent.com/95641633/147720219-0a73fd08-b5fa-4268-a5d8-908b3d41dd4d.png)

Both wild boar and Water deer used 150 sheets for testing(val.py). 
There are no Goral results in the matrix picture above. (Goral = 0 sheet)
but In the case of the Goral test, 132 videos were conducted( I checked it out with my own eyes.). and model predicted that the results were all Goral (model got the answer right).

![F1_curve](https://user-images.githubusercontent.com/95641633/147720229-24bd3f11-8854-4b43-ab54-3374605de2e6.png)

![P_curve](https://user-images.githubusercontent.com/95641633/147720245-d296427b-4891-4276-b4c4-35150fb47ab0.png)

![R_curve](https://user-images.githubusercontent.com/95641633/147720254-f7621a75-ffa2-4c5a-b38f-8e853fdc8f48.png)

![PR_curve](https://user-images.githubusercontent.com/95641633/147720260-60ba650b-1835-4941-b4fe-8bcc7229bfa8.png)

### Additional

* Wild boar and Water deer were taken pictures in similar places. Goral is a photo taken in a different place.
But all three were filmed in the wild.
* The data ratio is 7:2:1. ( train : val : test )
* I checked that it was well labeling.
* All the pictures were taken during the day.

**Comment by glenn-jocher** [comment]
@HeeHee100 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. It looks like your validation results are not showing one class because there are no examples of that class on the set you ran.

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/](https://github.com/ultralytics/yolov5/issues/2844#issuecomment-851338384)

Good luck 🍀 and let us know if you have any other questions!

**Comment by hee100** [comment]
Sorry. I didn't have enough explanation. I predicted test set instead of validation set with val.py.

In the case of the Goral test, 132 videos were inspected(I checked it myself). and model predicted that the results were all Goral (model got the answer right).

My problem is Deer and wild boar have a lot of FP each other. but the model predicts the Goral class perfectly.
As can be seen in the matrix above, model are not predict wild boars and deer to be Goral.
